logger/sdkLogger.go

```diff
@@ -0,0 +1,336 @@
+package logger
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+// SdkLoggerInterface Represents an interface for logging API requests and responses.
+type SdkLoggerInterface interface {
+	// LogRequest logs the details of an HTTP request.
+	LogRequest(request *http.Request)
+	// LogResponse logs the details of an HTTP response.
+	LogResponse(response *http.Response)
+}
+
+// SdkLogger represents implementation for SdkLoggerInterface, providing methods to log HTTP requests and responses.
+type SdkLogger struct {
+	loggingOptions LoggerConfiguration
+	logger         LoggerInterface
+}
+
+// NewSdkLogger Constructs a new instance of SdkLogger.
+func NewSdkLogger(loggingOpt LoggerConfiguration) *SdkLogger {
+	return &SdkLogger{
+		loggingOptions: loggingOpt,
+		logger:         loggingOpt.logger,
+	}
+}
+
+// LogRequest request Logs an HTTP request.
+func (a *SdkLogger) LogRequest(request *http.Request) {
+	var logLevel = a.loggingOptions.level
+	var contentTypeHeader = a._getContentType(request.Header)
+	var url string
+	if a.loggingOptions.request.includeQueryInPath {
+		url = request.RequestURI
+	} else {
+		url = a._removeQueryParams(request.RequestURI)
+	}
+
+	a.logger.Log(
+		logLevel,
+		fmt.Sprintf("Request %v %v %v", request.Method, url, contentTypeHeader),
+		map[string]any{
+			"method":      request.Method,
+			"url":         url,
+			"contentType": contentTypeHeader,
+		})
+
+	a._applyLogRequestOptions(logLevel, request)
+}
+
+// LogResponse Logs an HTTP response.
+func (a *SdkLogger) LogResponse(response *http.Response) {
+	var logLevel = a.loggingOptions.level
+	var contentTypeHeader = a._getContentType(response.Header)
+	var contentLengthHeader = a._getContentLength(response.Header)
+
+	a.logger.Log(
+		logLevel,
+		fmt.Sprintf("Response %v %v %v", response.StatusCode, contentLengthHeader, contentTypeHeader),
+		map[string]any{
+			"statusCode":    response.StatusCode,
+			"contentLength": contentLengthHeader,
+			"contentType":   contentTypeHeader,
+		},
+	)
+
+	a._applyLogResponseOptions(logLevel, response)
+}
+
+func (a *SdkLogger) _applyLogRequestOptions(level Level, request *http.Request) {
+	a._applyLogRequestHeaders(
+		level,
+		request,
+		a.loggingOptions.request,
+	)
+
+	a._applyLogRequestBody(level, request, a.loggingOptions.request)
+}
+
+func (a *SdkLogger) _applyLogRequestHeaders(
+	level Level,
+	request *http.Request,
+	logRequest RequestLoggerConfiguration) {
+
+	logHeaders := logRequest.headers
+	headersToInclude := logRequest.includeHeaders
+	headersToExclude := logRequest.excludeHeaders
+	headersToWhitelist := logRequest.whitelistHeaders
+
+	if logHeaders {
```

Comment:
`applyLogRequestHeaders` and `applyLogResponseHeaders` functions needs refactoring

They both are doing the same thing i.e getting a list of loggable headers from either request of response. You can simply create a GetLoggableHeaders(headers) function for the same thing

Reply:
Refactor these function to implement log with minimal code